---
title: "My Life In Sequences"
author: "Leighton Pritchard"
output: 
  revealjs::revealjs_presentation:
    theme: white
    center: true
    highlight: pygments
    transition: fade
    css: reveal.css
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## My life in sequences

<img src="images/comics/didnt_work_autobiography.jpg" height="450px" />
<img src="images/comics/what_i_planned.png" height="450px" />

A wonky path to bioinformatics

## A confession

<img src="images/comics/imposter_syndrome.jpg" height="450px" />

# The single most important influence on my academic career

## Informal education

<img src="images/1200px-Sinclair-ZX81.png" height="450px" />

## Informal education

<img src="images/comics/einstein_education.jpg" height="450px" />


# Formal Education

## BSc Forensic & Analytical Chemistry (1992-1996)

* Mostly chemistry (1yr industry), final year forensics
* Final year honours project: proteins and computers
* Sinclair Basic and Frontier: Elite II

<img src="images/bpti_sequence.png" height="300px" />
<img src="images/upgma_tree.png" height="300px" />
<img src="images/interchem_structure.png" height="300px" />

## Body Fluid Analysis (1996)

* Glasgow Royal Infirmary

<img src="images/peach_tea.jpg" height="450px" />

## Making the switch

<img src="images/comics/experimental_computational.jpg" height="450px" />

## PhD (1996-1999)

* Evolution of snake venom toxins (correlated mutations)
* Neural network model of protein evolution
* Drug active site discovery algorithm (still in use)

<img src="images/bpti_trace.png" height="250px" />
<img src="images/sid_scores.png" height="250px" />

[doi:10.1006/jmbi.1998.2437](https://dx.doi.org/10.1006/jmbi.1998.2437); 
[doi:10.1006/jtbi.1999.1043](https://dx.doi.org/10.1006/jtbi.1999.1043); 
[PMID:11579223](https://www.ncbi.nlm.nih.gov/pubmed/11579223); 
[PMID:12676977](https://www.ncbi.nlm.nih.gov/pubmed/12676977)

# First steps in academia

## First steps in academia

<img src="images/comics/phd031014s.gif" height="450px" />

## Postdocs (1999-2003)

* Systems biology: modelling yeast metabolism
* Directed evolution: improved lanthanide binding

<img src="images/glycolysis_equations.png" height="300px" />
<img src="images/glycolysis_domains.png" height="300px" />

[doi:10.1046/j.1432-1033.2002.03055.x](https://dx.doi.org/10.1046/j.1432-1033.2002.03055.x); 
[doi:10.1016/j.jtbi.2004.12.005](https://dx.doi.org/10.1016/j.jtbi.2004.12.005)

## Multidisciplinarity

<img src="images/comics/multidisciplinary.jpg" height="450px" />

# The SCRI/JHI Years

## The SCRI/JHI Years

<img src="images/comics/bioinformatics_keyboard.jpg" height="350px" />
<br></br>
<img src="images/SCRIlogoHighRes.png" height="100px" />
<img src="images/jhi_logo.png" height="100px" />

## Bioinformatician (2003-present)

* Research Institute
* Government funding, policy remits
* Ineligible for many usual funding sources (e.g. RCUK)
* Bioinformatician/Computational Biologist
* Not a clear postdoc/PI distinction
* BA Mathematics (Open University)

## Bioinformatician (2003-present)

* Bacterial genomics
* Plant-microbe interactions
* Oomycete genomics
* Microbial systems biology
* Microbial synthetic biology
* (occasionally plants, fungi, nematodes and viruses)

# Bacterial Genomics I

## 2003: Erwinia

* Arrived at SCRI part-way through sequencing *Erwinia carotovora* subsp. *atrosepticum*

<img src="images/cumulative_genomes_1.jpg" height="300px" />
<img src="images/cumulative_genomes_2.jpg" height="300px" />

([http://sulab.org/2013/06/sequenced-genomes‐per‐year/](h5p://sulab.org/2013/06/sequenced‐genomes‐per‐year/))

## Global pathogens

* Blackleg, stem-rot
* Quarantine pathogens

<img src="images/pathogen_ortho.png" height="350px" />
<img src="images/rot_pics1.png" height="350px" />

Toth *et al.* (2011) [doi:10.1111/j.1365-3059.2011.02427.x](https://dx.doi.org/10.1111/j.1365-3059.2011.02427.x)

## First enterobacterial plant pathogen genome
 
<img src="images/pba_pnas.png" height="300px" />

 [doi:10.1073/pnas.0402424101](https://dx.doi.org/doi:10.1073/pnas.0402424101)

## Acceptable in the 00s…

* 32-author single bacterial genome paper!
* £250,000 collaboration between SCRI, University of Cambridge, Wellcome Trust Sanger Institute
* All repeats and gaps bridged and sequenced directly
* A single complete high-quality 5Mbp circular chromosome
* 3 person-years' manual annotation

## TIL: Annotation

* The utility of genomes depends on annotation
* Annotation is **curation**, not cataloguing
* Automated annotation from curated data (e.g. with `Prokka`) now the only game in town
* ***But** you can't propagate what doesn't exist*
* Lots of genomes, few incentives to curate well: "many parents, but no-one wants to look after the children"
* Centralised data resources (e.g. PhytoPath) only a partial solution (funding?)

## TIL: Comparisons

* The real power of genomics is **comparative genomics**
* $\textrm{genome} \implies \textrm{heritable}$
* identification of functional elements, evolutionary processes and constraints
* **But** epigenetics, tissue differentiation, mesoscale, phenotype plasticity, …

<img src="images/genotype_protein_phenotype.png" height="200px" />

## Comparisons

* So we compared *Pba* to the 130 prokaryotic genomes available at the time…

<img src="images/pba_circular_rbh.png" height="350px" />
<img src="images/pba_linear_rbh.png" height="350px" />

[doi:10.1146/annurev.phyto.44.070505.143444](https://dx.doi.org/10.1146/annurev.phyto.44.070505.143444)

## Functional differences

* *Pba*-only: pathogenicity determinants; phage/IS elements
* *Pba* & environmental: pathogenicity determinants; surface proteins; regulatory proteins

<img src="images/pba_function_hist.png" height="400px" />

# Visualisation

## Visualisation

<img src="images/comics/how_scientists_see_data.jpg" height="350px" />

## GenomeDiagram

<img src="images/paper_gd.png" height="175px" />
<br></br>
<img src="images/gd1.png" height="250px" />
<img src="images/gd2.png" height="250px" />

[doi:10.1093/bioinformatics/btk021](https://dx.doi.org/10.1093/bioinformatics/btk021)

## `Biopython`/`KGML`

<img src="images/biopython.png" height="75px" />
<br></br>
<img src="images/kgml2.png" height="325px" />
<img src="images/kgml3.png" height="325px" />

[http://biopython.org](http://biopython.org)

## Art/Science

<img src="images/art_practice.jpg" height="300px" />
<img src="images/sciart2.png" height="300px" />
<img src="images/shemilt-fig3.jpg" height="300px" />

* Elaine Shemilt, exhibited Singapore, Dundee, London

[A Blueprint for Bacterial Life and Art](https://www.dundee.ac.uk/djcad/research/researchprojectscentresandgroups/blueprint/)

# Bacterial Genomics II

## Bacterial genomics II

<img src="images/comics/genomics_flying_cars.jpg" height="450px" />

## SPI-7 & *cfa*

* 11 horizontally-acquired islands: one similar to a *P. syringae* phytotoxin synthesis island (payload swapped)

<img src="images/spi7.png" height="400px" />

## SPI-7 & *cfa*

* Island present in *Pba*, some *Pcc*, no *Dickeya*
* *cfl* and *cfa7* Tn mutants showed reduced virulence (17dpi)

<img src="images/spi7_img.png" height="350px" />

Slawiak & Lojkowska (2009) [doi:10.1007/s10658-008-9418-7](10.1007/s10658-008-9418-7)

## 2013: Dickeya

* Genomics commodified (several services used)
* 25 genomes, no fanfare, minor publications

<img src="images/dickeya_ga1.png" height="150px" />
<img src="images/dickeya_ga2.png" height="150px" />

* 6 species, draft genomes, automated annotation

[doi:10.1128/genomeA.00978-13](https://dx.doi.org/10.1128/genomeA.00978-13)
[doi:10.1128/genomeA.00087-12](https://dx.doi.org/10.1128/genomeA.00087-12)

## Quarantine

* *Dickeya*: global pathogens

<img src="images/pathogen_maps.png" height="400px" />

## Diagnostic qPCR primers

* Genomes enable bulk design

<img src="images/design_stage_3.png" height="350px" />

[doi:/10.1371/journal.pone.0034498](https://dx.doi.org/10.1371/journal.pone.0034498)
<br></br>
[https://github.com/widdowquinn/find_differential_primers](https://github.com/widdowquinn/find_differential_primers)

## Tangled Taxonomy

* 3/4 *Dickeya* spp. misclassified at NCBI (MiSI: 18%)

<img src="images/dickeya_misclassified.png" height="400px" />

[doi:10.1111/j.1365-3059.2012.02678.x](https://dx.doi.org/10.1111/j.1365-3059.2012.02678.x)

## Nomenclature

* Old nomenclature/species: polyphasic, phenotypic
* Binomial nomenclature not designed for large amounts of genomic data, metadata curation (HGT of function…)
* Historical *Dickeya*  renaming: collections, databases not updated

<img src="images/sankey_taxonomy.png" height="300px" />

## Legislation

* A political issue: legislation/policy relies on binomial nomenclature

<img src="images/legislation.png" height="400px" />

## DDH

* DNA-DNA hybridisation the 'gold standard' for classification

<img src="images/ddh.png" height="400px" />

## ANI

* Average nucleotide identity (ANI) ≈ *in silico* DDH

<img src="images/ddh_anim.png" height="400px" />

## PYANI

* `Python` module and scripts

<img src="images/pyani_github_web.png" height="400px" />
<img src="images/SRE_ANIm.png" height="400px" />

## Dickeya

* Nine species-level groups (two novel)

<img src="images/Figure_2_figure_anim_pid_dickeya.png" height="500px" />

## Genomic classification

<img src="images/baltrus_tree.png" height="450px" />

Baltrus (2016) [doi:10.1016/j.tim.2016.02.004](https://dx.doi.org/j.tim.2016.02.004)

## ANIm Graphs

* Numeric matrices (e.g. ANIm scores) define graphs/networks

<img src="images/SRE_ANIm.png" height="400px" />
<img src="images/SRE_ANIm_graph.png" height="400px" />

## Graph decomposition

* Break down graphs (min. coverage, %identity)
* Identify unique *cliques*

<img src="images/thresholds.png" height="300px" />
<img src="images/threshold_zeros_kde.png" height="300px" />

* 'genus', 'species', 'clonal' groupings

[doi:10.6084/m9.figshare.4810867.v4](https://doi.org/10.6084/m9.figshare.4810867.v4)

## Reclassify genus

* 50%+ of genome aligns, ≈85% identity

<img src="images/sankey_genus.png" height="500px" />

## Overall classification

* Additional species reclassifications for SRE

<img src="images/sankey_gcs.png" height="500px" />

# Protecting woodlands

## Phyto-threats

<img src="images/project_flyer.png" height="350px" />

Forest Research, Centres for Ecology and Hydrology, University of Edinburgh, University of Worcester

## Phytophthora spp.

* *Phytopthora* spp. are devastating pathogens
* Crops routinely treated - woodlands aren't, but trees also susceptible to imported disease
* Sudden oak death; juniper (*P. austrocedri*), etc.
* Phyto-threats: evolutionary genomics to inform nursery practice
* Metabarcoding of *Phytophthora* communities in nurseries and surroundings
* Existing identification by ITS1, but tens of diverse ITS1 sequences per genome…

# Systems and Synthetic Biology

## SysBio and SynBio

<img src="images/comics/infinite_unfathomable_sadness.png" height="400px" />
<img src="images/genotype_protein_phenotype.png" height="250px" />

## Dickeya Metabolism

* Presence/absence of KEGG reactions
* Clues to host range and community function

<img src="images/dickeya_pathways.png" height="450px" />

## Flux-balance analysis

* Whole-organism flux, prediction of KO effects
* Substrate usage, pathway optimisation (SynBio)

<img src="images/fba_solution_space.png" height="450px" />

## Pathways and host range

* Pathways, function, correspond to *Dickeya* spp.

<img src="images/systems_dickeya_pathway_tree.png" height="325px" />
<img src="images/systems_dickeya_growth1.png" height="325px" />
<img src="images/systems_dickeya_growth2.png" height="325px" />

* Differences in substrate usage

## Food or Fuel?

* 1G: food crops; 2G: cellulosic crops; Ag waste?

<img src="images/275_cartoon_for_food_or_fuel_large.png" height="400px" />

* Maize stover, straw, sugarcane bagasse etc. 

## Dickeya for bioprocessing

* Plant Cell Wall Degrading Enzymes (PCWDEs)
* Enzymes engineered for bioethanol production, expressed in *E. coli*
* SynBio is a platform technology
* Engineered PCWDE libraries for SynBio?

<img src="images/logo_ibioic.png" height="75px" />
<img src="images/logo_uosta.png" height="75px" />

## Natural diversity

* Up to 75 CAZyme families in SRE
* Mine genomes for more variants

<img src="images/functional_space.png" height="400px" />

## Generating diversity

* Gene-shuffling/directed evolution
* Exploit/recombine natural diversity
* Obtain novel structures

<img src="images/generate_diversity.png" height="400px" />

## Protein sectors

* Correlated mutations (return of the PhD!)
* Decomposition of structure into sectors

<img src="images/correlated_conservation.png" height="350px" />
<img src="images/protein_sectors.png" height="350px" />

Halabi *et al.* (2009) [doi:10.1016/j.cell.2009.07.038](https://dx.doi.org/10.1016/j.cell.2009.07.038)

## Positional epistasis

* Context-dependence of function and mutation: epistasis
* Saturating substitution of specificity sites

<img src="images/epistasis.png" height="350px" />

* McLaughlin et al. (2012) [doi:10.1038/nature11500](https://dx.doi.org/10.1038/nature11500)

## Long way to go

<img src="images/comics/long_way_to_go.jpg" height="500px" />

# Conclusions

## Conclusions

<img src="images/comics/neil_degrasse_tyson_quote.jpeg" height="500px" />

## Microbial Agrogenomics

* Diagnostics & epidemic tracking by sequencing (MinION)
* SysBio: plant-microbe interactions, phytobiome (plant and associated community)
* Integration of models and datasets is still challenging
* SynBio: engineering new response modes into crops (resistance, N-fixation)

## As much as I know…

* Do work you enjoy
* Do it well
* Work with good people
* Read widely

## Acknowledgements

<img src="images/acknowledgements.png" height="500px" />