{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<img src=\"images/JHI_STRAP_Web.png\" style=\"width: 150px; float: right;\">\n",
    "# 02 - `UniProt`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Table of Contents\n",
    "\n",
    "1. [Introduction](#introduction)\n",
    "2. [The UniProt website](#uniprot)\n",
    "  1. [Using UniProtKB](#uniprotkb)\n",
    "  2. [Advanced Searches in UniProtKB](#uniprotkb_advanced)\n",
    "3. [Exercise 01](#ex01)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"introduction\"></a>\n",
    "## Introduction\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>`UniProt` is a comprehensive protein sequence and annotation resource, and is a consortium between the European Bioinformatics Institute (EMBL-EBI), the Swiss Institute of Bioinformatics (SIB), and the Protein Information Resource (PIR). `UniProt` provides a database that unifies several legacy databases, uncluding Swiss-Prot, TrEMBL, iProClass and the PIR-PSD. </b>\n",
    "</div>\n",
    "\n",
    "`UniProt` provides three key databases:\n",
    "\n",
    "* The UniProt Knowledgebase [(UniProtKB)](http://www.uniprot.org/help/uniprotkb)\n",
    "* The UniProt Reference Clusters [(UniRef)](http://www.uniprot.org/help/uniref)\n",
    "* The UniProt Archive [(UniParc)](http://www.uniprot.org/help/uniparc)\n",
    "\n",
    "`UniProtKB` is likely to be the database you use most frequently to find information on gene product/protein molecular function. It is the central hub of functional information on proteins, and collates functional annotations from many other databases, ontologies, and references. It keeps records of how annotations are derived (e.g. experimentally or computationally), and is divided into two sections: one contains manually-annotated records (`UniProtKB/Swiss-Prot`), and a second contains computational annotations that are waiting for manual curation (`UniProtKB/TrEMBL`).\n",
    "\n",
    "`UniRef` provides clustered sets of sequences from `UniProtKB` and `UniParc`. A number of clusterings at different stringencies are provided.\n",
    "\n",
    "`UniParc` is a comprehensive, non-redundant database that contains most of the publicly-available protein sequences from a range of sources.\n",
    "\n",
    "These databases can be queried in a number of ways, including:\n",
    "\n",
    "* At the `UniProt` website [http://www.uniprot.org/](http://www.uniprot.org/) in your web browser\n",
    "* Sending requests to the `UniProt` website, using a programming language\n",
    "\n",
    "### Resources\n",
    "* [`UniProt`](http://www.uniprot.org/)\n",
    "* [EMBL-EBI](http://www.ebi.ac.uk/)\n",
    "* [SIB](http://www.isb-sib.ch/)\n",
    "* [PIR](http://pir.georgetown.edu/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"uniprot\"></a>\n",
    "## The `UniProt` website\n",
    "\n",
    "* Open your browser and navigate to the [`UniProt` webpage](http://www.uniprot.org/). \n",
    "\n",
    "![UniProt landing page](images/06-01_landing.png)\n",
    "\n",
    "The landing page offers options for each of the three main databases: UniProtKB, UniRef, and UniParc. It also offers sets of [complete proteomes for a range of organisms](http://www.uniprot.org/proteomes/), and databases of proteins organised by supporting data, such as [literature](http://www.uniprot.org/citations/), [taxonomic classification](http://www.uniprot.org/citations/), and [subcellular location](http://www.uniprot.org/locations/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"uniprotkb\"></a>\n",
    "### Using `UniProtKB`\n",
    "\n",
    "* **Click on the `UniProtKB` link**. This will take you to the UniProtKB front page, with a summary of entries, and a number of links.\n",
    "\n",
    "![UniProtKB landing page](images/06-02_uniprotkb.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many records are in UniProtKB today?\n",
    "<li> How many of those records have been manually reviewed? What proportion of the total database has been manually reviewed?\n",
    "<li> Which organisms are most highly represented in the database, today?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* **Enter the word *Kitasatospora* in the search bar at the top of the page, and click on the `Search` button**.\n",
    "\n",
    "![UniProtKB Kitasatospora entries](images/06-03_kitasatospora.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many entries are returned?\n",
    "<li> How many of those entries have been manually reviewed? What proportion of the total is this?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Filtering Results\n",
    "* On the left of the page there's an option to filter `\"kitasatospora\"` as an organism or by taxonomy. **Click on the `organism` filter**.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many entries are returned?\n",
    "<li> How many of those entries have been manually reviewed?\n",
    "<li> How has the search term in the top bar changed? <b>NOTE: these search term changes will be useful for querying UniProt programmatically.</b>\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* At the top left of the page there's an option to filter only the manually reviewed entries. **Click on this filter**.\n",
    "\n",
    "![UniProtKB Kitasatospora reviewed entries](images/06-04_reviewed.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Inspecting an Entry\n",
    "\n",
    "* **Click on the link/accession for the topmost entry**.\n",
    "\n",
    "![UniProtKB reviewed entry header](images/06-05_evidence.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> What kind of evidence is there for protein function?\n",
    "<li> Kinetic information for this enzyme is drawn from which other database(s)?\n",
    "<li> Are any protein structures available for this enzyme?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* At the top of the page, there's a button marked `History`. Click on this button. A small window will open, with a link to `Previous versions`. **Click on this link**.\n",
    "\n",
    "![UniProtKB history window](images/06-06_history_link.png)\n",
    "\n",
    "![UniProtKB history](images/06-07_history_list.png)\n",
    "\n",
    "* **Click on the `Compare` button**.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> When was the last change made to this record?\n",
    "<li> What was the change?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* Compare some previous records to the current record (e.g. [this comparison](http://www.uniprot.org/uniprot/Q9AJE3?diff=true&version=30&version=44)).\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> What kinds of changes do you see?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"uniprotkb_advanced\"></a>\n",
    "### Advanced Searches in UniProtKB\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>At the top of the `UniProtKB` page you've probably noticed a drop-down button marked \"Advanced\". This lets you combine several search filters to conduct powerful searches, and hone in on the proteins most of interest to you in the UniProtKB database.</b>\n",
    "</div>\n",
    "\n",
    "In this section, you'll use the advanced searches to identify candidate human proteins that are found in the nucleus, and have been associated with some disease activity or function.\n",
    "\n",
    "* **Click on the UniProt logo to return to the landing page**\n",
    "* **Click on the \"Advanced\" drop-down to get the advanced searching interface**\n",
    "\n",
    "![UniProtKB advanced search](images/06-08_advanced.png)\n",
    "\n",
    "* **In the first field, select `Organism [OS]` with search term \"Human\"**. The dropdown will offer you several options as you type, but do not select them (you could have entered the organism \"Homo sapiens\" here, also).\n",
    "* **In the next field, keep `AND` on the left, and select `Subcellular location` with search term \"nucleus\"**. The dropdown will offer several options but, again, do not select them. At this point, allow any assertion method for the evidence code.\n",
    "\n",
    "![UniProtKB advanced search - first terms](images/06-09_search_1.png)\n",
    "\n",
    "* **Click on the plus sign (`+`) to get another search term field.**\n",
    "* **In the new field, keep `AND` on the left, and select `Pathology & Biotech` with class \"Disease\", and no search term**. At this point, allow any assertion method for the evidence code.\n",
    "\n",
    "![UniProtKB advanced search - second terms](images/06-10_search_2.png)\n",
    "\n",
    "* **Click on the magnifying glass to run the search.**\n",
    "\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many results do you see, today?\n",
    "<li> What are the contents of the search bar? **NOTE: this will be useful for programmatic queries, later.** \n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "#### Refine the search\n",
    "\n",
    "* **Click on the \"Advanced\" drop down.** You should see that the current search populates this dialogue box.\n",
    "* **Change the `Evidence` option for the `Pathology & Biotech` part of the search to manual \"Experimental\" evidence.**\n",
    "\n",
    "![UniProtKB advanced search - second terms](images/06-11_search_3.png)\n",
    "\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many results do you see, now?\n",
    "<li> What are the contents of the search bar? **NOTE: this will be useful for programmatic queries, later.** \n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* **Click on the \"Advanced\" drop down.** You should see that the current search populates this dialogue box.\n",
    "* **Change the `Term` option for the `Pathology & Biotech` part of the search to \"melanoma\".**\n",
    "\n",
    "![UniProtKB advanced search - second terms](images/06-12_search_4.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many results do you see, now?\n",
    "<li> How are these proteins associated with melanoma?\n",
    "<li> What amino acid modifications have been found for these proteins?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"download\"></a>\n",
    "### Downloading `UniProtKB` Search Results\n",
    "\n",
    "After the search above, you should be left with a small set of proteins that satisfy the following criteria:\n",
    "\n",
    "* They derive from *Homo sapiens*\n",
    "* They are annotated as being found in the nucleus (for which we allow any form of evidence)\n",
    "* They are associated with a disease process: melanoma, and there is manually-curated experimental evidence for this association\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>If we would like to download these records (or those from any other search), we have a number of options, which are obtained by clicking on the download button at the top of the search results.</b>\n",
    "</div>\n",
    "\n",
    "![UniProtKB download button](images/06-13_download.png)\n",
    "\n",
    "* You can download all your search results, or just those selected with checkboxes\n",
    "* Results can be downloaded compressed (`gzip`ped) or as raw records\n",
    "* Results can be downloaded as:\n",
    "  * sequence data (`FASTA`)\n",
    "  * tabular form (`Excel`, `tab-separated`)\n",
    "  * computer-readable (`XML`, `RDF`)\n",
    "  \n",
    "![UniProtKB download button](images/06-14_download_options.png)\n",
    "\n",
    "* **Download the search results as tab-separated, text, and FASTA format files**\n",
    "* **Inspect the contents of these files**\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How do the contents of these files differ\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<img src=\"images/exercise.png\" style=\"width: 100px; float: left;\">\n",
    "<a id=\"ex01\"></a>\n",
    "### Exercise 01 (15min)\n",
    "\n",
    "Using the `UniProtKB` search tools, can you find and download sets of proteins that satisfy the following requirements:\n",
    "\n",
    "<br></br>\n",
    "<div class=\"alert-danger\">\n",
    "<h4>Set 1</h4>\n",
    "<ul>\n",
    "<li> Derives from *Saccharomyces cerevisiae*\n",
    "<li> Is associated with a membrane\n",
    "<li> Is a transcriptional regulator\n",
    "</ul>\n",
    "<p></p>\n",
    "<h4>Set 2</h4>\n",
    "<ul>\n",
    "<li> Has a function associated with alginate\n",
    "<li> Is annotated with any biotechnological application\n",
    "</ul>\n",
    "<h4>Set 3</h4>\n",
    "<ul>\n",
    "<li> Any manual annotation associated with biofuel as a biotechnological application\n",
    "</ul>\n",
    "<h4>Set 4</h4>\n",
    "<ul>\n",
    "<li> Derives from mouse\n",
    "<li> Is an enzyme\n",
    "<li> Has not been manually reviewed\n",
    "<li> Is mentioned in a *Nature* publication\n",
    "<li> Is between 100 and 300aa in length\n",
    "</ul>\n",
    "</div>\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (SfAM)",
   "language": "python",
   "name": "python3_sfam"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
