{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<img src=\"images/JHI_STRAP_Web.png\" style=\"width: 150px; float: right;\">\n",
    "# 03 - `KEGG`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Table of Contents\n",
    "\n",
    "1. [Introduction](#introduction)\n",
    "2. [The `KEGG` website](#kegg)\n",
    "3. [`KEGG` GENOME](#genome)\n",
    "3. [`KEGG` GENES](#genes)\n",
    "3. [Exercise 01](#ex01)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"introduction\"></a>\n",
    "## Introduction\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>`KEGG` is an integrated resource of 16 databases, centred around a collection of biological pathway maps, covering metabolism, genetic information processing (regulation), environmental information processing (sensing and signalling), and cellular processes. It also covers drug metabolism, and provides a unique resource to integrate biological systems, genetic processes, and biochemical reactions.</b>\n",
    "</div>\n",
    "\n",
    "The pathway maps - 'wiring diagrams' of molecular interconversions - are hand-drawn, and form the core of `KEGG`. They integrate many molecular types, including proteins, RNA, primary and secondary metabolites, and chemical reactions. The maps are classified in a number of ways (metabolism, disease, and so on), each of which can be considered a conceptual 'view' onto the pathway data.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "In 2011, `KEGG`'s main data repository - the FTP site - moved to a subscription model, due to funding constraints. As a result, it is no longer possible to download freely the complete `KEGG` collection. However, access via the web browser, and for limited webservice queries, remains free.\n",
    "</div>\n",
    "\n",
    "`KEGG` can be queried in a number of ways, including:\n",
    "\n",
    "* At the `KEGG` website [http://www.genome.jp/kegg/pathway.html) in your web browser\n",
    "* Sending requests to the `KEGG` website, using a programming language\n",
    "\n",
    "### Resources\n",
    "* [`KEGG`](http://www.genome.jp/kegg/pathway.html)\n",
    "  1. [Metabolism](http://www.genome.jp/kegg/pathway.html#metabolism)\n",
    "  2. [Genetic Information Processing](http://www.genome.jp/kegg/pathway.html#genetic)\n",
    "  3. [Environmental Information Processing](http://www.genome.jp/kegg/pathway.html#environmental)\n",
    "  4. [Cellular Processes](http://www.genome.jp/kegg/pathway.html#cellular)\n",
    "  5. [Organismal Systems](http://www.genome.jp/kegg/pathway.html#organismal)\n",
    "  6. [Human Diseases](http://www.genome.jp/kegg/pathway.html#disease)\n",
    "  7. [Drug Development](http://www.genome.jp/kegg/pathway.html#drug)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"kegg\"></a>\n",
    "## The `KEGG` website\n",
    "\n",
    "* **Open your browser and navigate to the [`KEGG` webpage](http://www.genome.jp/kegg/pathway.html).** \n",
    "\n",
    "![KEGG landing page](images/08-01_KEGG_landing.png)\n",
    "\n",
    "The landing page offers a menu of choices: `PATHWAY`, `BRITE`, `MODULE` and so on (the main page shown is the `PATHWAY` page) , which each link to different databases within `KEGG`. For example:\n",
    "\n",
    "* [`BRITE`](http://www.genome.jp/kegg/brite.html): Functional hierarchies and binary relationships of biological entities\n",
    "* [`MODULE`](http://www.genome.jp/kegg/module.html): Functional units for annotating and interpreting genomes\n",
    "* [`KO`](http://www.genome.jp/kegg/ko.html): Linking genomes to pathways by ortholog annotation\n",
    "\n",
    "A full account of all the databases at `KEGG` is beyond the scope of this lesson.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"genome\"></a>\n",
    "## `KEGG` GENOME\n",
    "\n",
    "* [`KEGG` GENOME](http://www.genome.jp/kegg/genome.html): Organisms and ecosystems with genome sequence information\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>The `KEGG` databases contain information on a subset of sequenced organisms. These are described in the `KEGG` GENOME database, and each genome is identified by a unique three-letter code (though the code `map` is reserved for generic pathway maps).</b>\n",
    "</div>\n",
    "\n",
    "* **Click on the GENOME link in the KEGG menu bar**. This will take you to the `KEGG` GENOME database landing page.\n",
    "\n",
    "![KEGG GENOME landing page](images/08-02_kegg_genome.png)\n",
    "\n",
    "The landing page presents you with a search field.\n",
    "\n",
    "* **Enter the word \"Kitasatospora\" into the search field, and click on \"Go\"**\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many entries are returned?\n",
    "<li> What is the theoretical upper limit on the number of three-letter codes? What does this imply about the capacity of `KEGG` GENOME's naming scheme?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* **Click on the first link in the list of genomes.**\n",
    "\n",
    "![KEGG GENOME search result](images/08-03_kegg_genome_ksk.png)\n",
    "\n",
    "The `KEGG` entry for a genome typically links out to the GenBank record used for the `KEGG` entry, and also links internally to related genes and pathways in the other `KEGG` databases.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> Which `KEGG` databases are linked from the GENOME entry?\n",
    "<li> How many <i>Kitasatospora</i> genes are annotated?\n",
    "<li> How many `KEGG` GENES entries are linked from  <i>Kitasatospora</i> genes?\n",
    "<li> How many `KEGG` PATHWAY pathways include genes from <i>Kitasatospora</i>?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"genes\"></a>\n",
    "## `KEGG` GENES\n",
    "\n",
    "* [`KEGG` GENES](http://www.genome.jp/kegg/genes.html): Molecular building blocks of life in the genomic space\n",
    "\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>The `KEGG` GENES database contains the set of gene catalogues for all complete genomes in the `KEGG` GENOMES database.</b>\n",
    "</div>\n",
    "\n",
    "When genes are imported into the `KEGG` system, several analyses are applied, including assignment to `KO` (KEGG Orthology) groups that attempt to cluster genes that have similar function, on the basis of sequence similarity. These `KO` groups have their own `KEGG` KO database:\n",
    "\n",
    "* [`KEGG` KO](http://www.genome.jp/kegg/ko.html): KO Database of Molecular Functions\n",
    "\n",
    "* **Click on the menu link to the GENES database.**\n",
    "\n",
    "![KEGG GENES landing page](images/08-04_kegg_genes_landing.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>The `KEGG` GENES landing page presents three search fields, allowing you to search the GENES database, a specific organism, or to identify orthologues/paralogues/motifs and other connected information for a given gene.</b>\n",
    "</div>\n",
    "\n",
    "Keeping the *Kitasatospora* genome with code `ksk` as our example from above, we will search for genes whose annotation includes the phrase `xylulose`:\n",
    "\n",
    "* **Set `Organism` to `ksk` and enter `xylulose` in the search field, then click `Go`.**\n",
    "\n",
    "![KEGG GENES xylulose search](images/08-05_xylulose_search.png)\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many gene entries are returned?\n",
    "<li> How many different EC numbers are represented in the returned results?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "We will select one of the returned genes to focus on: `KSE_17560`.\n",
    "\n",
    "* **Click on the link for `KSE_17560`.**\n",
    "\n",
    "![KSE_17560 entry](images/08-06_KSE_17560.png)\n",
    "\n",
    "The gene record describes links to `KEGG` resources such as pathways, functional modules, and predicted orthologues, paralogues and gene clusters.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many `KEGG` PATHWAY pathways include this gene?\n",
    "<li> How many `KEGG` MODULE modules include this gene?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "The record also shows the amino acid and coding sequences (with optional upstream and downstream sequence) for the gene. Links to these sequences can be followed via the `AA seq` or `NT seq` buttons.\n",
    "\n",
    "* **Click on the AA seq button.** \n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How would you download this sequence?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"pathway\"></a>\n",
    "## `KEGG` PATHWAY\n",
    "\n",
    "* [`KEGG` PATHWAY](http://www.genome.jp/kegg/pathway.html): Wiring diagrams of molecular interactions, reactions, and relations\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>The PATHWAY database is very much the heart of `KEGG`. It enables the mapping of individual elements of a genome (or several genomes) in the context of large-scale dynamic systems, such as metabolism and other cellular processes. It is a hugely valuable resource for interpreting genes and gene products in a cellular, systems-level context.</b>\n",
    "</div>\n",
    "\n",
    "![KEGG PATHWAY landing page](images/08-01_landing.png)\n",
    "\n",
    "The PATHWAY landing page gives a single search field, allowing a free-text search of the complete PATHWAY database. Note that the \"Organism\" selected by default is `map`, which is the generic reference map.\n",
    "\n",
    "* **Enter \"terpenoid\" in the search field and click \"Go\".**\n",
    "\n",
    "![KEGG PATHWAY search result](images/08-07_pathway_result.png)\n",
    "\n",
    "The search result lists pathways that match the free text search, and gives a short account of each.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many pathways are returned for this search?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* **Click on `map00900`.**\n",
    "\n",
    "![KEGG map00900](images/08-08_map00900.png)\n",
    "\n",
    "The map shown is a reference map, and is interactive: \n",
    "\n",
    "* clicking on a rectangular node will take you to an entry in the `KEGG` ORTHOLOGY (KO) database\n",
    "* clicking on a circular node will take you to an entry in the `KEGG` COMPOUND database\n",
    "* clicking on a red node will take you to the connected entry in the `KEGG` PATHWAY database\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-danger\">\n",
    "<b>Currently the map has no colour highlighting, as it is a reference map. When you select a pathway map related to a specific organism, the rectangular nodes are highlighted if there is a gene annotated to have that function in the genome.</b>\n",
    "</div>\n",
    "\n",
    "* **Return to the `KEGG` PATHWAY search page.**\n",
    "* **Enter `ksk` in the \"Organism\" field, and search again for \"terpenoid\".**\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many pathways are returned for this search?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "Note that the search result pathways now have the prefix `ksk`. This is how `KEGG` allows us to index pathway maps directly: the three-letter code identifies the organism, and the five digit number identifies the pathway. Combining the two as `<organism><pathway>` means we can directly construct the pathway map ID for any combination of organism and pathway.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> The three-letter code for *Homo sapiens* is `hsa`. What would be the ID for the human terpenoid backbone biosynthesis pathway?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "* **Click on the `ksk00900` map**. This will open the terpenoid backbone biosynthesis map for *Kitasatospora setae*.\n",
    "\n",
    "![KEGG ksk00900](images/08-09_ksk00900.png)\n",
    "\n",
    "The `ksk` map is identical in structure to the `map` map, except that some of the rectangular boxes representing chemical interconversions have now been shaded green. This indicates that there is a gene annotated in the organism whose product is expected to be able to perform the biochemical interconversion indicated on the map.\n",
    "\n",
    "* **Click on the green box with `2.2.1.7`.**\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> How many genes are linked from the box marked `2.2.1.7`?\n",
    "</ol>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<a id=\"special\"></a>\n",
    "### \"Special\" maps\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "<b>Most of `KEGG`'s pathway maps are drawn like those above, graphs where black lines as edges link rectangular (reaction) and circular (compound) nodes. There are currently four overview maps that are rendered quite differently:</b>\n",
    "</div>\n",
    "\n",
    "* [`map01100`](http://www.genome.jp/kegg-bin/show_pathway?map01100): Metabolic pathways overview\n",
    "* [`map01110`](http://www.genome.jp/kegg-bin/show_pathway?map01110): Biosynthesis of secondary metabolites\n",
    "* [`map01120`](http://www.genome.jp/kegg-bin/show_pathway?map01120): Microbial metabolism in diverse environments\n",
    "* [`map01130`](http://www.genome.jp/kegg-bin/show_pathway?map01130): Biosynthesis of antibiotics\n",
    "\n",
    "Following the links above, these are interactive. The individual pathways can be highlighted by selecting the checkboxes on the left hand side, and clicking on the circles will link out to compounds, while clicking on the graph edges will link out to reactions.\n",
    "\n",
    "![KEGG map of antibioitic biosynthesis, with erythromycin synthesis highlighted](images/08-10_antibiotics.png)\n",
    "\n",
    "* **Return to the main `KEGG` PATHWAYS database page.**\n",
    "* **Click on the link to \"Biosynthesis of sntibiotics\" in section 1.0 (Global and overview maps).** This will bring up the interactive map.\n",
    "* **Select `M00774` Erythromycin biosynthesis** on the left-hand side. Note that the erythromycin pathway is highlighted.\n",
    "* **Click on the \"Reference pathway\" drop-down menu. Scroll down to Kitasatospora setae, and select that organism.**  **HINT:** if you type \"kitasat\" with the drop-down selected, that will take you to the *Kitasatospora* option quickly.\n",
    "* **Click on \"Go\".**\n",
    "\n",
    "After a short delay, you should see that many of the edges and nodes in the image are greyed out. These pathways are not annotated in *K. setae*. The pathways that remain coloured are, however, annotated as present in that organism.\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-warning\">\n",
    "<b>QUESTIONS</b>\n",
    "<ol>\n",
    "<li> Which antibiotic synthesis pathways/modules from `KEGG` are annotated as being present in *Kitasatospora setae*?\n",
    "</ol>\n",
    "</div>\n",
    "\n",
    "![KEGG PATHWAY antibiotic synthesis](images/08-11_antibiotic.png)\n",
    "\n",
    "To download a copy of the image you have generated with your selections, click on the `Image (png) file` link at the top of the page."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "<img src=\"images/exercise.png\" style=\"width: 100px; float: left;\">\n",
    "<a id=\"ex01\"></a>\n",
    "### Exercise 01 (15min)\n",
    "\n",
    "The `UniProt` record `Q05655` describes a human protein kinase. Using `KEGG`, can you discover:\n",
    "\n",
    "<br></br>\n",
    "<div class=\"alert-danger\">\n",
    "<ul>\n",
    "<li> What is the record for this gene in the `KEGG` GENE database?\n",
    "<li> Which pathways is gene product associated with?\n",
    "<li> What diseases is this enzyme associated with?\n",
    "<li> What are the substrates and products of this enzyme, and what is the `KEGG` REACTION database entry for the reaction?\n",
    "<li> One of the pathways this enzyme is associated with is the chemokine signalling pathway. In the map for this pathway, what are the downstream pathways from this enzyme?\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### SOLUTION - EXERCISE 01\n",
    "\n",
    "* [hsa:5580](http://www.genome.jp/dbget-bin/www_bget?hsa:5580)\n",
    "* [12 pathways](http://www.genome.jp/dbget-bin/get_linkdb?-t+pathway+hsa:5580)\n",
    "* [Type II diabetes](http://www.genome.jp/kegg-bin/show_pathway?hsa04930) and [insulin resistance](http://www.genome.jp/kegg-bin/show_pathway?hsa04931)\n",
    "* [ATP](http://www.genome.jp/dbget-bin/www_bget?cpd:C00002) and [protein](http://www.genome.jp/dbget-bin/www_bget?cpd:C00017); [ADP](http://www.genome.jp/dbget-bin/www_bget?cpd:C00008) and [phosphoprotein](http://www.genome.jp/dbget-bin/www_bget?cpd:C00562); [reaction](http://www.genome.jp/dbget-bin/www_bget?rn:R00162)\n",
    "* Degranulation, Chemotaxis, NO induction, ROS production"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (SfAM)",
   "language": "python",
   "name": "python3_sfam"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
